metagraph/src/graph/annotated_graph_algorithm.cpp

```diff
@@ -15,249 +12,416 @@ namespace graph {
 using mtg::common::logger;
 
 typedef AnnotatedDBG::node_index node_index;
-typedef AnnotatedDBG::row_index row_index;
 typedef AnnotatedDBG::Annotator::Label Label;
 
+typedef std::function<size_t()> LabelCountCallback;
+
+constexpr std::memory_order MO_RELAXED = std::memory_order_relaxed;
 
-std::unique_ptr<bitmap_vector>
-mask_nodes_by_unitig(const DeBruijnGraph &graph,
-                     const KeepUnitigPath &keep_unitig,
-                     size_t num_threads) {
-    sdsl::bit_vector unitig_mask(graph.max_index() + 1, false);
 
-    const bool atomic = num_threads > 1;
+/**
+ * Return an int_vector<>, bit_vector pair, of lengths anno_graph.get_graph().max_index() * 2
+ * and anno_graph.get_graph().max_index(), respectively.
+ * For an index i, the int_vector at indices 2*i and 2*i + 1 represent the
+ * number of labels in labels_in and labels_out which the k-mer of index i is
+ * annotated with, respectively. The width of the int_vector<> is computed to be
+ * wide enough to contain counts up to num_labels.
+ * The returned bit_vector is a k-mer mask indicating those k-mers which are annotated
+ * with at least one in-label or out-label.
+ */
+std::pair<sdsl::int_vector<>, sdsl::bit_vector>
+construct_diff_label_count_vector(const AnnotatedDBG &anno_graph,
+                                  const tsl::hopscotch_set<Label> &labels_in,
+                                  const tsl::hopscotch_set<Label> &labels_out,
+                                  size_t num_labels,
+                                  size_t num_threads);
+
+// Regions of a graph mask which should be kept (i.e., masked in)
+typedef std::vector<std::pair<size_t, size_t>> Intervals;
+
+// Returns a vector of kept regions given a unitig and its corresponding path
+typedef std::function<Intervals(const std::string&, const std::vector<node_index>&)> GetKeptIntervals;
+
+// Returns true if a node is kept (i.e., masked in)
+typedef std::function<bool(node_index)> KeepNode;
```

Comment:
this only makes thing confusing. Better write `std::function<bool(node_index)>` -- it's short and descriptive.